### DISCUSSION TITLE
Extending the exception handling?

### DISCUSSION THREAD

**Discussion by elfring** [curl/curl#9425] [Ideas]
:eyes: I took another look at a few usages of [the macro “failf”](https://github.com/curl/curl/blob/6fa89fa893cb8d10b0e4c6c77f88f9bf5d9d4411/lib/sendf.h#L48 "Macro definition").
[The function “Curl_failf”](https://github.com/curl/curl/blob/6fa89fa893cb8d10b0e4c6c77f88f9bf5d9d4411/lib/sendf.c#L259 "Handling of failure messages") is called accordingly.

I came along [the need for more detailed reports after selected errors were detected](https://progress.opensuse.org/issues/116024 "How to fix error reports “No route to host” for mirror service implementations?").
I imagine that additional error information can be determined on demand for better explanation of potentially unexpected software run time situations.
How do you think about the possibility to configure customised callback functions for further exception handling? :thinking:

**Comment by dfandrich** [comment]
libcurl already has a callback to receive error messages.  What kind of API are you proposing?

**Reply by bagder** [reply]
Yups, there is also [CURLOPT_ERRORBUFFER](https://curl.se/libcurl/c/CURLOPT_ERRORBUFFER.html) if you want the failf() error message alone.

**Reply by elfring** [reply]
>libcurl already has a callback to receive error messages.

* Can further data processing be performed by the means of other data structures (according to an exceptional situation) than such a “message”?
* Would you like to add any links for the software documentation?


>What kind of API are you proposing?

I find it occasionally helpful to perform additional customised failure analysis.
Extra code would be passed as a pointer to a callback function.
This function will probably need two pointer parameters.
* Varying libcurl context (or handle)
* Custom data (`void*`)


Will the error analysis trigger further software development challenges? 🤔

**Comment by dfandrich** [comment]
So far, I haven't seen you propose anything that isn't already supported by 
CURLOPT_DEBUGFUNCTION and CURLOPT_DEBUGDATA. You'll have to be more specific.

**Reply by elfring** [reply]
I got the impression that [the “debug function”](https://curl.se/libcurl/c/CURLOPT_DEBUGFUNCTION.html "debug callback") would affect all libcurl function calls.
I became more interested in the determination of additional error data for selected “exceptions”.
* How would you like to represent attempts for finding helpful background information according to an error hint like “[No route to host](https://progress.opensuse.org/issues/116024 "How to fix error reports “No route to host” for mirror service implementations?")”? :thinking:
* :eyes: Are you looking then for support of more detailed data structures than a varying string (`char *`)?

**Reply by elfring** [reply]
Would you occasionally get into the mood to [inspect network connection information in more detail](https://github.com/curl/curl/blob/472f1cbe7e65a8807878e6309926ddb67c0770ff/lib/urldata.h#L887 "Structure “connectdata”") and perform further data processing for desired error analyses? :thinking:

**Reply by bagder** [reply]
> [the “debug function”](https://curl.se/libcurl/c/CURLOPT_DEBUGFUNCTION.html) would affect all libcurl function calls

Sure, but in practice does it really matter? It is the network transfer process that generates 99.7% (made up number) of the data for that callback.

> How would you like to represent attempts for finding helpful background information according to an error hint like “No route to host”?

I don't think anything can be added to curl to help with a no route to host error. That's a general IP issue that is outside of what curl does and knows about.

> Would you occasionally get into the mood to [inspect network connection information in more detail](https://github.com/curl/curl/blob/472f1cbe7e65a8807878e6309926ddb67c0770ff/lib/urldata.h#L887)

Yes. I have done it more than once. I think our problem to understand what you're asking for is that we have worked with and solved these kinds of problems for decades already using the means already provided, so it would be helpful if you could be more specific exactly what you would get from your proposed API that you can't get now.

**Reply by elfring** [reply]
>I don't think anything can be added to curl to help with a no route to host error.

💭 I hope that this software library can become more helpful for such known error situations.


>That's a general IP issue that is outside of what curl does and knows about.

Will the chances evolve any further to improve corresponding exception handling?


>… that we have worked with and solved these kinds of problems for decades already using the means already provided, …

Are failed network connections represented by higher level data structures than an error code like “`EHOSTUNREACH`”?

**Reply by bagder** [reply]
> I hope that this software library can become more helpful for such known error situations.

You're more than welcome to explain to us how you think we can make it so!

> Will the chances evolve any further to improve corresponding exception handling?

I don't understand this question.

> Are failed network connections represented by higher level data structures than an error code like “EHOSTUNREACH”?

Sorry, I don't understand this question either.

**Reply by elfring** [reply]
>You're more than welcome to explain to us how you think we can make it so!

Desirable software evolution will depend on supported abstraction and service levels.

How much are error codes only converted to error message strings (instead of selecting a more sophisticated errror response)? :thinking:

**Reply by elfring** [reply]
>I don't understand this question.

How do you think about to take another look at an information source like “[United States Patent 11,356,352 (from 2022-07-06): Identifying reachability of network-connected devices](https://patft.uspto.gov/netacgi/nph-Parser?Sect1=PTO1&Sect2=HITOFF&d=PALL&p=1&u=%2Fnetahtml%2FPTO%2Fsrchnum.htm&r=1&f=G&l=50&s1=11,356,352.PN.&OS=PN/11,356,352&RS=PN/11,356,352 "United States Patent: 11,356,352")”? :thinking:

**Reply by bagder** [reply]
I can't take *another* look on something I've never seen before. If you cannot explain your point without referring to a page full of  gibberish, then I'm not sure it is a good idea.

**Reply by elfring** [reply]
>If you cannot explain your point without referring to a page full of gibberish,

The mentioned patent describes also some use cases where you can get concerned about the (un)reachability of known network services.
💭 Are you involved in “prior art” accordingly?


>then I'm not sure it is a good idea.

Would you like to configure error reporting in more desirable ways? :thinking:

**Reply by bagder** [reply]
I will just wait until I see a clear proposal for something.

**Reply by elfring** [reply]
💭 Will development interests grow for configuring the applied error handling in more detail?

**Reply by elfring** [reply]
Another wording approach:
💭 I propose to replace [the following source code](https://github.com/curl/curl/blob/ed6e0febe0a9e341d3998caff0962e0b76b06b7f/lib/connect.c#L1072 "Curl_is_connected(): Connection failure") by [a call to a customisable function](https://en.wikipedia.org/wiki/Callback_%28computer_programming%29 "Description for the usage of callback functions").

```c
    failf(data, "Failed to connect to %s port %u after "
          "%" CURL_FORMAT_TIMEDIFF_T " ms: %s",
          hostname, conn->port,
          Curl_timediff(now, data->progress.t_startsingle),
#ifdef ENABLE_QUIC
          (conn->transport == TRNSPRT_QUIC) ?
          curl_easy_strerror(result) :
#endif
          Curl_strerror(error, buffer, sizeof(buffer)));
```


:thinking:
* How do you think about to apply other error handling approaches at similar places?
* Would you become interested to work with the determined error code directly instead of generating only a varying error message?

**Reply by bagder** [reply]
How would that be different to the callback it already calls?

**Reply by elfring** [reply]
The corresponding callback function would provide customised error/exception handling after a connection failure was detected within [the implementation of the function “Curl_is_connected”](https://github.com/curl/curl/blob/ed6e0febe0a9e341d3998caff0962e0b76b06b7f/lib/connect.c#L863 "Update candidate").

**Comment by dfandrich** [comment]
It sounds like you're asking for a way to distinguish the particular 
error that generates that failf() call. Is that right? Could another way 
of accomplishing what you're asking for be having a much more 
fine-grained set of error codes, so that the application can distinguish 
the exact failure path of each error?

**Reply by elfring** [reply]
>It sounds like you're asking for a way to distinguish the particular error that generates that failf() call.

💭 My imaginations are evolving also into such a direction.


>Could another way of accomplishing what you're asking for be having a much more fine-grained set of error codes,

I find that better error information can generally be helpful for the handling of related use cases.


>so that the application can distinguish the exact failure path of each error?

The programming language “C” provides some known design challenges as they are discussed also for various software libraries according to more sophisticated exception handling. :thinking:
